psi4/driver/driver.py

```diff
@@ -1938,17 +1938,111 @@ def fchk(wfn, filename):
 
     :examples:
 
     >>> # [1] FCHK file for DFT calculation
     >>> E, wfn = energy('b3lyp', return_wfn=True)
     >>> fchk(wfn, 'mycalc.fchk')
 
+    >>> # [2] FCHK file for correlated densities
+    >>> E, wfn = gradient('ccsd', return_wfn=True)
+    >>> fchk(wfn, 'mycalc.fchk')
+
     """
+    #NOTE: This function should be overhauled once all modules set Ca/Cb, Da/Db in the same manner.
+
+    # QC_MODULE overview for FCHK
+    # OCC: (includes both occ/dfocc modules
+    #  wfn.D() = SCF
+    #  wfn.C() = SCF
+    #  refwfn.D() = SCF
+    # -> not allowed, graceful exit
+    # DCT:
+    #  wfn.D() = SCF
+    #  wnf.C() = DCT
+    #  refwfn.D() = SCF
+    # -> allowed, but FCHK entries are switched around
+    # CCENERGY
+    #  wfn.D() = CC
+    #  wfn.C() = SCF 
+    #  refwfn.D() = CC
+    # -> allowed, will recompute SCF density from wfn.C()
+    # DFMP2:
+    #  wfn.D() = MP2
+    #  wnf.C() = NO(MP2)
+    #  refwfn.D() = SCF 
+    # -> allowed, will take refwfn.D() for SCF density
+
+    not_working_correctly = ['OCC']
+    qc_module = core.get_global_option('QC_MODULE')
+    if (qc_module in not_working_correctly[:]):
+        core.print_out(f"QC_MODULE {qc_module} is currently not supported by the FCHK writer.")
+        return None
+
     fw = core.FCHKWriter(wfn)
+
+    # At this point we don't the the method name, so we try to search for it.
+    # idea: get the method from the variable matching closely the 'current energy'
+    varlist = core.variables()
+    current = varlist['CURRENT ENERGY']
+    del varlist['CURRENT ENERGY']
+
+    # delete problematic entries
+    delete_list = ['CURRENT ENERGY', 'CURRENT GRADIENT', 'CURRENT REFERENCE ENERGY']
+    for key in delete_list[:]:
+        if key in varlist:
+            del varlist[key]
+
+    # find closest match, skip arrays
+    for (key, val) in varlist.items():
+        if hasattr(val, '__len__'): continue
+        if (np.isclose(val, current)):
+            method = key
+            break
+
+    # The labels this functions knows. Extend if needed.
+    labels = {
+    "HF" :" SCF Density",
+    "DFT":" SCF Density",
+    "CC" :" CC Density",
+    "MP2":" MP2 Density",
+    "MP3":" MP3 Density",
+    "MP4":" MP4 Density",
+    "CI" :" CI Density",
+    "DCT":" DCT Density",
+    }
+    fchk_label= [val for (key,val) in labels.items() if key in method][0]
+    fw.set_postscf_density_label(fchk_label)
+
+    # Sometimes we have the SCF density on the reference_wavefunction and
+    # othertimes we need to recompute it. Only affects correlated densities.
+    if (qc_module in ['DFMP2']):
+        print('False')
+        fw.recompute_scf_density(False)
+    else:
+        fw.recompute_scf_density(True)
+
+    if 'DCT' in fchk_label:
+        # The Ca/Cb are the optimized orbitals from DCT. 
+        warn = " !! FCHK: In DCT calculations the SCF Density and DCT density entries are switched. !!"
+        core.print_out(warn)
```

Comment:
Obligatory comment that I'm not happy about this, but fine. As discussed in Slack, the C and D fields need to be standardized.

Reply:
I don't insist on enabling DCT. It's more a "one might as well since it's there".

Reply:
Rewording: "I'm not happy that we need such a system of hacks to get DCT enabled at all."